managed/services/inventory/inventory.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package inventory
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/AlekSi/pointer"
+	prom "github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm/managed/models"
+)
+
+const (
+	cancelTime                  = 3 * time.Second
+	serviceEnabled      float64 = 1
+	prometheusNamespace         = "pmm_managed"
+	prometheusSubsystem         = "inventory"
+)
+
+var (
+	mAgentsDesc = prom.NewDesc(
+		prom.BuildFQName(prometheusNamespace, prometheusSubsystem, "agents"),
+		"The current information about agent",
+		[]string{"agent_type", "service_id", "node_id", "pmm_agent_id", "disabled", "version"},
+		nil)
+	mNodesDesc = prom.NewDesc(
+		prom.BuildFQName(prometheusNamespace, prometheusSubsystem, "nodes"),
+		"The current information about node",
+		[]string{"node_type", "node_name", "container_name"},
+		nil)
+	mServicesDesc = prom.NewDesc(
+		prom.BuildFQName(prometheusNamespace, prometheusSubsystem, "services"),
+		"The current information about service",
+		[]string{"service_type", "node_id"},
+		nil)
+)
+
+type Inventory struct {
+	db *reform.DB
+	r  agentsRegistry
+}
+
+func NewInventory(db *reform.DB, r agentsRegistry) *Inventory {
+	i := &Inventory{
+		db: db,
+		r:  r,
+	}
+	return i
+}
+
+func (i *Inventory) Describe(chan<- *prom.Desc) {}
```

Comment:
Just a question because I really don't remember how it works internally.
If describe doesn't return anything, Collect is still going to be called once or twice?
I remember we had a problem with MongoDB exporter, Prometheus was calling DescribeByCollect, resulting in calling Collect twice per cycle.

Reply:
What should be returned in `Describe`?

Reply:
The list of metrics.

Reply:
Why we don't send descriptors in Describe?

Reply:
But metrics builded here
https://github.com/percona/pmm/pull/1854/files#diff-e9b7475ea5592bfc2b7287010a7920f4ed8b3d3c632d19d3b34fac7b4a53ce48R131
For any item of items - new metrics 
Because each of metrics have unique labels (unique name of agent, service, etc)

Reply:
@percona-csalguero
yes, it can be called 2 times in case use this one for example `prom.DescribeByCollect` in Descibe.
internally it works this way:
```
func DescribeByCollect(c Collector, descs chan<- *Desc) {
	metrics := make(chan Metric)
	go func() {
		c.Collect(metrics)
		close(metrics)
	}()
	for m := range metrics {
		descs <- m.Desc()
	}
}
```
i.e. it calls collect and resends these to desc.
it is not actually 2 collections. Describe called 1 time, on start.
---
@xgrommx I am not following your explanation.
what don't we use prom.DescribeByCollect?
of course we can do
```
	ch <- mAgentsDesc
	ch <- mNodesDesc
	ch <- mServicesDesc
```
but in this case we are not checking if we can collect metrics, with actual collection we see potential issues on startup, not first scrape.